### ISSUE TITLE
Producing outlines images in training data

### ISSUE THREAD

**Issue by emilyzhang393** [Shreeyak/cleargrasp#21]
The global optimization module which predicts the depth uses occlusion boundaries as an input. To help the edge prediction model learn better, we added an extra class of contact edges. 
The outlines are predicted through semantic segmentation, where each pixel is assigned a class. There are 2 types of edges: occlusion boundaries (which indicate a large differential on the depth) and the contact edges (edges in contact with other surfaces, and hence do not have the large differential).
In the paper, we visualize the 3 classes (background, occlusion, contact) through the colors red, green, blue.

_Originally posted by @Shreeyak in https://github.com/Shreeyak/cleargrasp/issues/17#issuecomment-766655544_

Hello I am creating my own training data and I am able to produce segmentation masks for my images but am unsure on how to produce the outlines images. Could you help me here?

**Comment by MMehdiMousavi** [comment]
@emilyzhang393 I have been studying this pipeline for about 6 months. I have my own custom-dataset creation system and Ive been able to reproduce all of the ground-truth passes. As far as I understand it (and asked from the authors) the occlusion outline is simply 2 different outlines, one is the entire outline from the objects, and one is where the Z values are lower than a certain amount, which causes wierd artifacts like i mentioned in https://github.com/Shreeyak/cleargrasp/issues/17 .

If you're still working on this, send me a DM or something. I can help you with the virtual data you need.

**Comment by Shreeyak** [comment]
> Hello I am creating my own training data and I am able to produce segmentation masks for my images but am unsure on how to produce the outlines images. Could you help me here?

Take a gradient of the segmentation masks to get the outline :).
As @MMehdiMousavi mentioned, do take a look at https://github.com/Shreeyak/cleargrasp/issues/17 . 
@MMehdiMousavi SuperCaustics looks amazing. Nice work!

**Comment by 1075976432** [comment]
> > Hello I am creating my own training data and I am able to produce segmentation masks for my images but am unsure on how to produce the outlines images. Could you help me here?
> 
> Take a gradient of the segmentation masks to get the outline :).
> @emilyzhang393 as mentioned, do take a look at #17 .
> @MMehdiMousavi SuperCaustics looks amazing. Nice work!

I'm doing my own real world dataset and has no idea how to make normal data from rgb and depth image. Could you give some tips for that?

**Comment by Shreeyak** [comment]
@1075976432 To calculate surface normals from a depth image, you'll first have to create a point cloud from the depth image. From that, you can calculate the surface normal for each point from the neighboring points.
Use a library such as Open3D to do both. Ref: http://www.open3d.org/docs/0.7.0/tutorial/Basic/pointcloud.html#vertex-normal-estimation

**Comment by Shreeyak** [comment]
Update: Found the script used for creating the outlines. Have added it to the repo here: https://github.com/Shreeyak/cleargrasp/blob/master/z-ignore-scripts-helper/data_processing_script.py
